packages/components/src/components/SquiggleViewer/ValueWithContextViewer.tsx

```diff
@@ -143,6 +155,12 @@ export const ValueWithContextViewer: FC<Props> = ({
   // In that case, the output would look broken (empty).
   const isOpen = !collapsible || !itemState.collapsed;
 
+  useEffect(() => {
+    if (isFocused && !isRoot && ref.current) {
+      focusOnHeader(ref.current);
+    }
+  }, []);
```

Comment:
This part seems hackish (both the dependencies warning and the way you have to expose `focusOnHeader` function).

Suggestion: instead of registering DOM elements in `ItemStore`, register custom imperative handles similar to ones that are used in React's `useImperativeHandle`.

Steps:
- export `ItemHandle` type from this class, with all necessary callbacks: `forceUpdate`, `focusOnHeader`, `scrollIntoView`
- change `useRegisterAsItemViewer` params to `(path: SqValuePath, handle: ItemHandle): void`
- move `useForceUpdate` here, and call `useRegisterAsItemViewer(path, { ...all callbacks }` here
- implement `focusOnHeader` by registering a separate `headerRef` in this component; then you won't need `element.querySelector("header")`

---

Now, about this `useEffect`... I assume your intended semantics is "when the component is mounted for the first time".

I'd suggest using a ref callback that registers the node in a normal `useRef` and also calls `focus()` when it's first initialized.

Something like this:
```typescript
const headerRef = useRef<HTMLHeaderElement | null>(null);

const focusOnHeader = () => headerRef.current?.focus();

return (
  ...
  <header
      ref={(node) => {
      headerRef.current = node;
      focusOnHeader();
    }}
  >
  ...
);
```

(I'm not 100% sure React won't call `ref` again on each render if its identity is unstable; is that's a problem, you'll have to move that lambda code to `useCallback`.)

Reply:
Thanks! Agreed it's kind of sketchy/hackish now, wasn't sure about an alternative though.

> Now, about this useEffect... I assume your intended semantics is "when the component is mounted for the first time".

Yep

Reply:
> Suggestion: instead of registering DOM elements in ItemStore, register custom imperative handles similar to ones that are used in React's useImperativeHandle.

Yep, I considered that, but wasn't sure if it was worth it. Will give it a try.

Reply:
I did most of that, except for that last bit. We can't set ``headerRef`` manually like that, and it's not officially set until the end of the function, so I don't think this technique will work. 

I think ``useEffect`` is the recommended way of doing this actually. That said, I'll mess with it to try other solutions.